File: pwa/src/domain/injury/load-adjustment-policy.ts
/**
 * load-adjustment-policy.ts
 * Phase-specific multipliers for exercise load, sets, and reps.
 * Ported from Swift LoadAdjustmentPolicy.swift.
 */

import { RecoveryPhase, ExerciseValue } from '../types';
import { swiftRound } from '../types';

export interface LoadMultipliers {
  load: number;
  sets: number;
  reps: number;
}

/**
 * Returns load/sets/reps multipliers for the given recovery phase.
 * Matches Swift LoadAdjustmentPolicy.multipliers(for:).
 */
export function getLoadMultipliers(phase: RecoveryPhase): LoadMultipliers {
  switch (phase) {
    case 'acute':
      return { load: 0.0, sets: 0.0, reps: 0.0 };
    case 'rehab':
      return { load: 0.30, sets: 0.50, reps: 0.70 };
    case 'lightLoad':
      return { load: 0.50, sets: 0.75, reps: 0.85 };
    case 'returnToPlay':
      return { load: 0.80, sets: 0.90, reps: 1.0 };
    case 'resolved':
      return { load: 1.0, sets: 1.0, reps: 1.0 };
  }
}

/**
 * Applies a multiplier to an ExerciseValue.
 * - fixed: rounds result, clamps to min 1
 * - range: rounds both ends, clamps to min 1
 * - amrap/text: pass through unchanged
 * Matches Swift LoadAdjustmentPolicy.adjustValue(_:multiplier:).
 */
export function adjustExerciseValue(ev: ExerciseValue, multiplier: number): ExerciseValue {
  switch (ev.kind) {
    case 'fixed':
      return { kind: 'fixed', value: Math.max(1, swiftRound(ev.value * multiplier)) };
    case 'range':
      return {
        kind: 'range',
        lo: Math.max(1, swiftRound(ev.lo * multiplier)),
        hi: Math.max(1, swiftRound(ev.hi * multiplier)),
      };
    case 'amrap':
      return ev;
    case 'text':
      return ev;
  }
}

/**
 * Applies the load multiplier to a percent1RM value.
 * Returns null when load is undefined (matching Swift Optional.none).
 * Matches Swift LoadAdjustmentPolicy.adjustLoad(_:multiplier:).
 */
export function adjustLoad(load: number | undefined, multiplier: number): number | null {
  if (load === undefined || load === null) return null;
  return load * multiplier;
}

File: SundeeFundee/Domain/LoadAdjustmentPolicy.swift
import Foundation

/// Phase-specific multipliers for exercise load, sets, and reps.
/// Applied to replacement exercises during injury adaptation.
enum LoadAdjustmentPolicy {

    struct Multipliers {
        let load: Double
        let sets: Double
        let reps: Double
    }

    /// Returns multipliers for a given recovery phase.
    static func multipliers(for phase: RecoveryPhase) -> Multipliers {
        switch phase {
        case .acute:
            return Multipliers(load: 0.0, sets: 0.0, reps: 0.0)
        case .rehab:
            return Multipliers(load: 0.30, sets: 0.50, reps: 0.70)
        case .lightLoad:
            return Multipliers(load: 0.50, sets: 0.75, reps: 0.85)
        case .returnToPlay:
            return Multipliers(load: 0.80, sets: 0.90, reps: 1.0)
        case .resolved:
            return Multipliers(load: 1.0, sets: 1.0, reps: 1.0)
        }
    }

    /// Applies multipliers to an exercise value (sets or reps).
    static func adjustValue(_ value: ExerciseValue, multiplier: Double) -> ExerciseValue {
        switch value {
        case .fixed(let n):
            return .fixed(max(1, Int(round(Double(n) * multiplier))))
        case .range(let lo, let hi):
            return .range(max(1, Int(round(Double(lo) * multiplier))), max(1, Int(round(Double(hi) * multiplier))))
        case .amrap, .text:
            return value
        }
    }

    /// Applies load multiplier to percent1RM.
    static func adjustLoad(_ percent1RM: Double?, multiplier: Double) -> Double? {
        guard let pct = percent1RM else { return nil }
        return pct * multiplier
    }

    /// Applies full multiplier set to a ProgramExercise.
    static func applyMultipliers(_ exercise: ProgramExercise, phase: RecoveryPhase) -> ProgramExercise {
        let m = multipliers(for: phase)
        guard m.load < 1.0 || m.sets < 1.0 || m.reps < 1.0 else { return exercise }
        return ProgramExercise(
            exercise: exercise.exercise,
            variant: exercise.variant,
            sets: adjustValue(exercise.sets, multiplier: m.sets),
            reps: adjustValue(exercise.reps, multiplier: m.reps),
            percent1RM: adjustLoad(exercise.percent1RM, multiplier: m.load),
            restMinutes: exercise.restMinutes,
            notes: exercise.notes,
            bodyweightOnly: exercise.bodyweightOnly
        )
    }
}
